R/language_level_df.R

```diff
@@ -0,0 +1,145 @@
+
+#' Reduce dialects and other duplicates which have the same glottcode in dataframe to one per language, and give it the glottocode of the parent that is at level language. Requires a cldf-value table and language-table. Simplifies combination of dataset.
+#'
+#' @param ValueTable data-frame, long format. ValueTable from cldf.
+#' @param method character vector, choice between "singular_least_missing_data", "combine_random", "singular_random". combine_random = combine all datapoints for all the dialects/duplicates and if there is more than one datapoint for a given feature/word/variable choose randomly between them, singular_random = choose randomly between the dialects/duplicates, singular_least_missing_data = choose the dialect/duplicate which has the most datapoints.
+#' @param LanguageTable data-frame of a cldf LanguageTable. Needs to have columns Glottocode and Language_ID or Language_level_ID. If not otherwise specified, it fetches the LanguageTable from glottolog-cldf online on GitHub (v4.8).
+#' #' @param drop_question_marks logical vector indicating wether to turn ? into missing values (NA) or not.
+#' @return data-frame of ValueTable without duplicates
+#' @export
+
+# TODO SJG: can we fix the indentation here?
+# Hedvig response: Sure, to me that's a aesthtics thing, we can do it however you'd like.
+
+
+language_level_df <- function(ValueTable = NULL,
+                              method = c( "singular_least_missing_data", "combine_random", "singular_random"),
+                              drop_question_marks = TRUE,
+                              LanguageTable = "https://github.com/glottolog/glottolog-cldf/raw/v4.8/cldf/languages.csv"){
+
+    if(length(method) != 1 & !(method %in% c( "singular_least_missing_data",
+                                              "combine_random", "singular_random"))){
+        stop("Method of merging is not defined.")
+    }
+
+    ### WRANGLING LANGUAGE TABLE
+    # ValueTable = readr::read_csv("tests/testthat/fixtures/testdata/values.csv")
+
+    # TODO SJG: should add some checks here to make sure we have ALL the columns we need
+    #  or we stop()
+    # Hedvig response: Sure.
+
+    if (!all(c('Language_ID', 'Parameter_ID', 'Value') %in% colnames(ValueTable))) {
+        stop("Invalid table format - ValueTable needs to have columns Language_ID/Parameter_ID/Value")
+    }
+
+if(LanguageTable == "https://github.com/glottolog/glottolog-cldf/raw/v4.8/cldf/languages.csv"){
+LanguageTable <- read.delim(LanguageTable, sep = ",")
+}
+
+# The language level ID column is named different in different, CLDF datasets.
+# setting them to the same
```

Comment:
I don't really understand this bit. AFAICT the function expects the LanguageTable from the glottolog-cldf dataset. Which other CLDF datasets with different column names would make sense as input here?

Reply:
It can also take a LanguageTable from some other cldf-dataset. 

This function can work on several different kinds of cldf-datasets where language-levelling is relevant. I've suggested before it could go in rcldf instead of rgrambank

Reply:
I set it to read glottolog 4.8 from github just as a default fallback.

Reply:
Ah, ok, I see that the relevant information is also available from the Grambank LanguageTable (to which it was copied from Glottolog for convenience). But different "levels" of languoids and relations between these via `Language_level_ID` aren't standardized in CLDF, but just Glottolog-specific ways to encode dialect/language classifications.

So I'd either require Glottolog's LanguageTable to be passed into this function, or just have one dataset (or `Grambank`) argument, and then use ValueTable and LanguageTable from the same dataset in this function.

Reply:
I think the solution here depends on if this function continues to live in rgrambank or is moved to rcldf.

Reply:
I agree with robert here -- this function needs to know three things:
1. a language table from a cldf dataset
2. a values table from a cldf dataset.
3. a glottolog dataset.

The current implementation relies on grambank only columns (Language_level_ID) which is ok, but it would be nice to make it more general (which isn't too far off but will require pulling this out of glottolog). 

More problematically,  it relies on a hardcoded table from glottolog. We should decouple this because glottolog evolves, and we don't want to be updating this every single time glottolog releases a new version. The package here has  a function to easily load_glottolog() so we can make users use that e.g. 

```r
df <- language_level_df(df, glottolog=load_glottolog())
```

I'll merge this for now and fix this myself.

Reply:
c.f. #17